src/librustc_mir/interpret/eval_context.rs

```diff
@@ -743,9 +743,22 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
 
             Discriminant(ref place) => {
                 let ty = self.place_ty(place);
+                let layout = self.layout_of(ty)?;
                 let place = self.eval_place(place)?;
                 let discr_val = self.read_discriminant_value(place, ty)?;
-                if let ty::TyAdt(adt_def, _) = ty.sty {
+                match layout.variants {
+                    layout::Variants::Single { index } => {
+                        assert_eq!(discr_val, index as u128);
+                    }
+                    layout::Variants::Tagged { ref variants, .. } |
+                    layout::Variants::NicheFilling { ref variants, .. } => {
+                        if variants.get(discr_val as usize).is_none() {
```

Comment:
Discriminants and variant indices don't match. You also need to adjust `read_discriminant_value` to return the variant index (so a usize) and adjust all calls to it to stop doing weird magic. In case of tagged encoding you always need to iterate like the old code did, but we should be doing so inside `read_discriminant_value`

Reply:
I just noticed it too